rule.py

```diff
@@ -1,100 +1,254 @@
+        else : return False 
 
-    def soldier_passed_river_upper_half(self, old_pos, new_pos):
-        pass 
+    def you_shall_not_pass(self, old_pos, new_pos):
+        '''
+        Determine if piece is in the group of elephant , advisor, general and apply its special rule of not leaving palace or country respectively
+        '''
 
-    def soldier_passed_river_lower_half(self, old_pos, new_pos):
-        pass
+        imperial_guard = ['elephant', 'advisor', 'general']
+        the_one = self.get_piece(old_pos).type 
+        if the_one in imperial_guard: 
+            the_origin_coords = the_one.his_pos[-1]            
+            if the_one == imperial_guard[0]: 
+                if self.is_in_lower_half(the_origin_coords): 
+                    return self.is_in_lower_half(new_pos)
+                elif self.is_in_upper_half(the_origin_coords): 
+                    return self.is_in_upper_half(new_pos)
+                else: 
+                    raise ValueError('The Elephant is travelling oversea!!! MovRuleSpecial_1: Chesspiece is not detected in board')
+            elif the_one == imperial_guard[1]: 
+                if self.is_in_northern_palace(the_origin_coords):
+                    return self.is_in_northern_palace(new_pos)
+                elif self.is_in_southern_palace(the_origin_coords):
+                    return self.is_in_southern_palace(new_pos)
+                else: 
+                    raise ValueError('Advisor not detected in the palace!!! MovRuleSpecial_2: Chesspiece is not detected in both palaces')
+            elif the_one == imperial_guard[2]: 
+                if self.is_in_southern_palace(the_origin_coords):
+                    return self.is_in_southern_palace(new_pos)
+                elif self.is_in_northern_palace(the_origin_coords): 
+                    return self.is_in_northern_palace(new_pos)
+                else: 
+                    raise ValueError('This is bad, the General is missing in the palace!!! MovRuleSpecial_3: Chesspiece is not detected in both palaces')
+            else: 
+                raise ValueError('What is the world is this thing? Identify thyself!! MovRuleSpecial_4: Chesspiece does not belong to the special group')
 
-    def cannon(self, old_pos, new_pos):
+    def to_be_or_not_to_be(self, old_pos, new_pos): 
+        '''
+        Determine if the soldier has passed the river and apply moving rule accordingly 
+        '''
         pass 
 
-    def chariot(self, old_pos, new_pos):
-        pass 
-    
-    def horse(self, old_pos, new_pos):
-        pass
-    
-    def elephant(self, old_pos, new_pos):
-        pass 
+    def is_interfered(self, old_pos, new_pos):
+        the_one = ['horse', 'elephant']
+        if self.get_piece(old_pos).type in the_one:  
+            if self.get_piece(old_pos).type == the_one[0]: 
+                return self.board_map_2d[new_pos.y][new_pos.x] != self._IS_EMPTY and self.is_interfered_horse(old_pos, new_pos)
```

Comment:
Either one of the case would cause an interference: an piece in new pos or a piece blocking the road. So, it should be an 'or' not 'and'